src/self-managed-debug/src/system_catalog_dumper.rs

```diff
@@ -615,57 +615,231 @@ const RELATIONS: &[Relation] = &[
+static SELECT_CLUSTER_REPLICAS_QUERY: &str = "SELECT c.name as cluster_name, cr.name as replica_name FROM mz_clusters AS c JOIN mz_cluster_replicas AS cr ON c.id = cr.cluster_id;";
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct ClusterReplica {
-    // TODO (debug_tool1): Use this when scraping introspection relations
-    pub _cluster_name: String,
-    pub _replica_name: String,
+    pub cluster_name: String,
+    pub replica_name: String,
+}
+
+impl Default for ClusterReplica {
+    fn default() -> Self {
+        Self {
+            cluster_name: "mz_catalog_server".to_string(),
+            replica_name: "r1".to_string(),
+        }
+    }
+}
+
+impl fmt::Display for ClusterReplica {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}.{}", self.cluster_name, self.replica_name)
+    }
 }
 
 pub struct SystemCatalogDumper<'n> {
     context: &'n Context,
-    pg_client: Arc<PgClient>,
-    // TODO (debug_tool1): Use this when scraping introspection relations
-    _cluster_replicas: Vec<ClusterReplica>,
+    pg_client: Arc<Mutex<PgClient>>,
+    pg_tls: MakeTlsConnector,
+    cluster_replicas: Vec<ClusterReplica>,
     _pg_conn_handle: JoinHandle<Result<(), tokio_postgres::Error>>,
 }
 
+pub async fn create_postgres_connection(
+    connection_string: &str,
+) -> Result<
+    (
+        PgClient,
+        Connection<Socket, TlsStream<Socket>>,
+        MakeTlsConnector,
+    ),
+    anyhow::Error,
+> {
+    let mut pg_config = PgConfig::from_str(connection_string)?;
+    pg_config.connect_timeout(PG_CONNECTION_TIMEOUT);
+    let tls = make_tls(&pg_config)?;
+    let (pg_client, pg_conn) = retry::Retry::default()
+        .max_duration(PG_CONNECTION_TIMEOUT)
+        .retry_async_canceling(|_| {
+            let pg_config = pg_config.clone();
+            let tls = tls.clone();
+            async move { pg_config.connect(tls).await.map_err(|e| anyhow::anyhow!(e)) }
+        })
+        .await?;
+
+    Ok((pg_client, pg_conn, tls))
+}
+
+pub async fn copy_relation_to_csv(
+    transaction: &Transaction<'_>,
+    file_path_name: PathBuf,
+    column_names: &mut Vec<String>,
+    relation_name: &str,
+) -> Result<(), anyhow::Error> {
+    let mut file = tokio::fs::File::create(&file_path_name).await?;
+
+    file.write_all((column_names.join(",") + "\n").as_bytes())
+        .await?;
+
+    // Stream data rows to CSV
+    let copy_query = format!(
+        "COPY (SELECT * FROM {}) TO STDOUT WITH (FORMAT CSV)",
+        relation_name
+    );
+
+    let copy_stream = transaction
+        .copy_out(&copy_query)
+        .await
+        .context(format!("Failed to COPY TO for {}", relation_name))?
+        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e));
+    let copy_stream = std::pin::pin!(copy_stream);
+    let mut reader = StreamReader::new(copy_stream);
+    tokio::io::copy(&mut reader, &mut file).await?;
+
```

Comment:
Here we should call `file.sync_all().await?` so we make sure the data is fsync-ed to disk